Notes/Proposals/Explicit-and-implicit-partial-type-inference.md

```diff
@@ -0,0 +1,332 @@
+What we could use instead of creating a new class is an alias representing that type.
+However current aliases are very strict to be used for that.
+We could extend them to support generic parameters to solve the issue.
+
+Other scenarios contain a specification of all type arguments when declaring generic types. 
+We feel, that the way is not ideal because of the unnecessary code that the programmer has to write. 
+See the example below. 
+We don't have to specify the second type argument, it can be inferred from the constructor. 
+However, the first argument cannot be omitted because we want to specify the precision.
+
+```csharp
+using System.Collections.Generic;
+
+Calculation<float, object> calculation = new (new List<object>());
+//or
+var calculation = new Calculation<float, object>(new List<object>());
+
+public class Calculation<TPrecission, TEntity> {
+    public Calculation(List<TEntity> entities) {} 
+}
+```
+
+The current generic method type inference situation works as an "all or nothing" principle. 
+If the compiler is not able to infer command call type arguments, a user has to specify all of them. 
+This requirement can be noisy and unnecessary in scenarios where some type arguments can be inferred by the compiler based on other dependencies.
+
+We can take an example from the `System.Linq` library. 
+When we use the `Select` method, the first argument can be inferred from a target. 
+However, the second argument doesn't have to be obvious from the context. 
+We can see it in the example, where we want to specify the precision. 
+The better way would be to skip obvious type parameters (`int` here) and specify just ambiguous ones(`float` here).
+
+```csharp
+using System.Collections.Generic;
+using System.Linq;
+
+var scores = new List<int>() {1,2,3};
+// Error: IEnumerable<float> preciseScores = scores.Select(i => i);
+IEnumerable<float> preciseScores = scores.Select<int, float>(i => i); // OK
+```
+
+## Scope
+
+Partial type inference can be solved in various ways. We can inspire from default and named arguments where we don't have to write all arguments. 
+Together with generic aliases, we will call it **Implicit partial type inference**. 
+
+Also, there are situations, where we want just to specify the arity of desired generic parameter or specify a parameter that is not possible to infer from the context but let the compiler infer the remaining arguments. 
+This part aims to suggest a way how to specify which positional arguments should be inferred by the compiler. 
+We will call it **Explicit partial type inference**.
+
+## Detailed design
+
+### Implicit partial type inference
+
+Implicit partial type inference is fully inspired by default and named arguments of argument list of function members with slight modification including bringing a new concept of `this` keyword and combination with *Explicit partial type inference*.
+
+#### Default type parameters
```

Comment:
I think for the proposal to be most successful in an environment like LDM, it will be best to focus it on the following:

1. for generic method calls, allow a type argument list which contains a combination of *actual types* and *placeholders* like `_` in `M<int, _>("a")`. Detail how the type argument inference process is different in these cases.
2. for object-creation expressions, allow a generic type being constructed to contain *placeholders*. e.g. `new MyComplexType<_>(arg)`, and define the type argument inference process that occurs.
3. (stretch goal) for local variable declarations and casts, permit types with placeholders like `List<_> local = expr;` or `(Complex<_, _>)expr` and define the process by which we decide the type of the resulting variable or expression respectively.

I'm not actually sure (3) can work. So, I guess it could be worth exploring :P. I do think that a proposal that just contains details on how to accomplish (1) and (2) would be worthwhile on its own.

It may be good to mention some of the other functionality as *possible expansions*, and describe why this proposal won't limit us from doing any potentially desirable expansions in the future.

However, I wouldn't actually include the proposal to do default type arguments, named type arguments, or generic aliases in this proposal. Although any interdependency between them must be understood, *whether to do* each one needs to be considered independently to some extent.

I will try and do a more thorough review pass tomorrow and provide more feedback where it may be useful.

Reply:
Great, Thanks! 
I think how to continue. I will split the other features into standalone proposals and describe possible interdependencies and etc. Also I will copy the inference algorithm description from the C# specification and modify it to work with `_`. And I will look at the rest of your comments.

Reply:
What we often do is link to appropriate section of csharpstandard, e.g. https://github.com/dotnet/csharpstandard/blob/draft-v7/standard/expressions.md#1163-type-inference.

Then we can limit the amount of spec content in the proposal to just paragraphs where we're adding or changing something. Those paragraphs can be pasted in a blockquote (a line beginning with `>`). Changes can be denoted with **bold** for new stuff and ~~strikethrough~~ for removed stuff.